dali/python/nvidia/dali/pickling.py

```diff
@@ -0,0 +1,144 @@
+    basic_def = (fun.__name__, fun.__qualname__, code, fun.__closure__)
+    fun_context = {
+        'global_refs': cl_vars.globals,
+        'defaults': fun.__defaults__,
+        'kwdefaults': fun.__kwdefaults__
+    }
+    return function_unpickle, basic_def, fun_context, None, None, set_funcion_state
+
+def module_unpickle(name, origin, submodule_search_locations):
+    if name in sys.modules:
+        return sys.modules[name]
+    spec = importlib.util.spec_from_file_location(
+        name, origin,
+        submodule_search_locations=submodule_search_locations)
+    module = importlib.util.module_from_spec(spec)
+    sys.modules[name] = module
+    spec.loader.exec_module(module)
+    return module
+
+def module_reducer(module):
+    spec = module.__spec__
+    return module_unpickle, (spec.name, spec.origin, spec.submodule_search_locations)
+
+def set_cell_state(cell, state):
+    cell.cell_contents = state['cell_contents']
+
+def cell_unpickle():
+    return types.CellType(None)
+
+def cell_reducer(cell):
+    return (cell_unpickle, tuple(), {'cell_contents': cell.cell_contents}, None, None, set_cell_state)
+
+
+class DaliCallbackPickler(pickle.Pickler):
+
+    def reducer_override(self, obj):
+        if inspect.ismodule(obj):
+            return module_reducer(obj)
+        if isinstance(obj, types.CellType):
+            return cell_reducer(obj)
+        if inspect.isfunction(obj):
+            if isinstance(obj, type(dummy_lambda)) and obj.__name__ == dummy_lambda.__name__ or \
+                    getattr(obj, '_dali_pickle_by_value', False):
+                return function_by_value_reducer(obj)
+            if '<locals>' in obj.__qualname__:
+                try:
+                    pickle.dumps(obj)
+                except AttributeError as e:
+                    if "Can't pickle local object" in str(e):
+                        return function_by_value_reducer(obj)
+        return NotImplemented
+
+
+def dumps(obj, protocol=None, **kwargs):
+    f = io.BytesIO()
+    DaliCallbackPickler(f, protocol, **kwargs).dump(obj)
+    return f.getvalue()
+
+
+loads = pickle.loads
+
+
+class CustomPickler:
+
+    @classmethod
+    def create(cls, py_callback_pickler):
+        if py_callback_pickler is None or isinstance(py_callback_pickler, cls):
+            return py_callback_pickler
+        if hasattr(py_callback_pickler, 'dumps') and hasattr(py_callback_pickler, 'loads'):
+            return cls.of_reducer(py_callback_pickler)
+        if isinstance(py_callback_pickler, (tuple, list)):
+            params = [None] * 3
+            for i, item in enumerate(py_callback_pickler):
+                params[i] = item
+            reducer, kwargs_dumps, kwargs_loads = params
+            return cls.of_reducer(reducer, kwargs_dumps, kwargs_loads)
+        raise ValueError("Unsupported py_callback_pickler value provided.")
+
+    @classmethod
+    def of_reducer(cls, reducer, dumps_kwargs=None, loads_kwargs=None):
```

Comment:
This is a weird name. Is it required by some API? What does this `of` stand for?

Reply:
Nah, it's not required by any API, I meant create_instance_of_this_class_from_reducer and for some reason used of rather than from.